### DISCUSSION TITLE
Unexpected behavior with get_env() defaults

### DISCUSSION THREAD

**Discussion by rsyring** [jdx/mise#2519] [General]
I have a situation where I conditionally need a Terraform provider.  But, the provider always checks for it's env vars and raises an error if not found, even when it will not be used.  Some people using Terraform may not have the authorization to use that provider but should still be able to use terraform.  

The provider can be silenced by providing placeholder credentials:

```
TAILSCALE_OAUTH_CLIENT_ID = "placeholder"
TAILSCALE_OAUTH_CLIENT_SECRET = "placeholder"
```

I use a different tool that can read secrets from 1Pass to set those environment variables when I really need them: `env-config tailscale`

My goal is to set the placeholders by default for anyone working on the project so that terraform "just works."  But also be able to set them for real when Tailscale tokens are actually needed by terraform.  

I assumed this would work in my `mise.toml`:

```toml
TAILSCALE_OAUTH_CLIENT_ID = "{{ get_env(name='_TAILSCALE_OAUTH_CLIENT_ID', default='placeholder') }}"
TAILSCALE_OAUTH_CLIENT_SECRET = "{{ get_env(name='_TAILSCALE_OAUTH_CLIENT_SECRET', default='placeholder') }}"
```

And I've config `env-config tailscale` to set the underscore versions:

```
 ❯ env-config tailscale
Clearing:
     No conigured vars present to clear.
Profiles active: tailscale
Setting:
    _TAILSCALE_OAUTH_CLIENT_ID: op://team-devops-privileged/tailscale-oauth-client/username
    _TAILSCALE_OAUTH_CLIENT_SECRET: op://team-devops-privileged/tailscale-oauth-client/password
```

What I've observed with this setup:

1. After running `env-config tailscale` it takes about a minute for mise to show the env vars to the new values instead of the placeholders.
2. When removing the underscore versions, mise doesn't ~~seem to ever~~ notice for about 10 mins.  It continues to show the values set by env-config.
3. Using `mise cache clear` does not affect either of the above

I'm using mise's integration with the Fish shell in case that matters.

Thought about filing an issue but figured I'd ask first to make sure this is not expected behavior for some reason.

---

Edit: might actually be something other than time triggering the change but I'm not sure what.  It seems I can trigger an update by changing directories but I think I've seen it change previously without doing that.

**Comment by jdx** [comment]
some logging would be helpful to figure out where it is getting stuck, I would try setting MISE_TRACE=1 but that may produce too much garbage to be useful. There is also a way to log to a file but I forget the syntax off-hand, if you can't find that in the docs I'll see if I can get you some instructions.

**Comment by Marukome0743** [comment]
I investigated this again. The current behavior is intentional: environment variables read by config templates are not automatically tracked as reactive inputs.

PR #11570 explored automatically invalidating `hook-env` and environment caches when values accessed through `get_env()` changed. However, that approach was not accepted because it introduced substantial session/cache invalidation complexity and would have made `get_env()` reactive while the equivalent `env.NAME` access remained non-reactive.

When an external command changes variables used by a mise template, it should explicitly force mise to reevaluate the environment. For Fish, this can be done with:

```fish
env-config tailscale
mise hook-env --force -s fish | source
```

The same forced refresh should be run after clearing the underscore-prefixed variables so that the placeholder defaults are restored.

Users who enable `hook_env.cache_ttl`, `hook_env.chpwd_only`, or environment caching should otherwise expect changes from external environment-mutating tools to be reflected only after the relevant cache or hook condition is invalidated. `mise cache clear` does not directly force the active shell’s `hook-env` fast path to rerender the environment.

So the recommended solution here is to integrate the forced `hook-env` refresh into `env-config`, or run it immediately after changing those variables, rather than relying on mise to detect the changes automatically.

*AI-assisted — Tool: Codex; model: OpenAI/GPT-5; version: unavailable.*
